intellij/src/saros/intellij/editor/EditorManager.java

```diff
@@ -123,109 +121,192 @@ private void execEditorActivity(EditorActivity editorActivity) {
+            setLocalDocumentModificationHandlersEnabled(true);
+          }
+        }
+
+        /**
+         * Adjusts all current annotations for the given file according to the given text edit
+         * offsets. Adds a contribution annotation if the text edit added new text.
          *
-         * @param user the user for the given operation
-         * @param file the file for the given operation
-         * @param editor the editor for the given file
-         * @param operations the received operation
-         * @param calculationEditor an editor for the file; this might be a background editor (see
-         *     {@link BackgroundEditorPool}), so it must not be used for annotation purposes
+         * @param user the user that caused the text edit
+         * @param file the file that was modified
+         * @param editor the editor for the file
+         * @param start the start offset of the text edit
+         * @param oldEnd the end offset of the replaced text
+         * @param newEnd the end offset of the new text
+         * @see AnnotationManager#moveAnnotationsAfterDeletion(IFile, int, int)
+         * @see AnnotationManager#moveAnnotationsAfterAddition(IFile, int, int)
          */
         private void adjustAnnotationsAfterEdit(
             @NotNull User user,
             @NotNull IFile file,
             @Nullable Editor editor,
-            @NotNull Operation operations,
-            @NotNull Editor calculationEditor) {
-
-          operations
-              .getTextOperations()
-              .forEach(
-                  textOperation -> {
-                    TextPosition startPosition = textOperation.getStartPosition();
-                    TextPosition endPosition = textOperation.getEndPosition();
-
-                    int start = EditorAPI.calculateOffset(calculationEditor, startPosition);
-                    int end = EditorAPI.calculateOffset(calculationEditor, endPosition);
-
-                    if (textOperation instanceof InsertOperation) {
-                      if (editor == null) {
-                        annotationManager.moveAnnotationsAfterAddition(file, start, end);
-                      }
+            int start,
+            int oldEnd,
+            int newEnd) {
 
-                      annotationManager.addContributionAnnotation(user, file, start, end, editor);
+          if (oldEnd != start && editor == null) {
+            annotationManager.moveAnnotationsAfterDeletion(file, start, oldEnd);
+          }
 
-                    } else if (textOperation instanceof DeleteOperation && editor == null) {
+          if (newEnd != start) {
+            if (editor == null) {
+              annotationManager.moveAnnotationsAfterAddition(file, start, newEnd);
+            }
 
-                      annotationManager.moveAnnotationsAfterDeletion(file, start, end);
-                    }
-                  });
+            annotationManager.addContributionAnnotation(user, file, start, newEnd, editor);
```

Comment:
If `editor == null` is true, the `addContributionAnnotation` is also called. I would expect that an annotation is always assigned to an editor. The editor parameter is @Nullable, but what happens with an annotation that is created for `editor == null`?

Reply:
This is explained in the javadoc of [AbstractEditorAnnotation](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AbstractEditorAnnotation.java#L25).

The thought behind allowing the editor to be `null` is that there isn't always an open editor for the resource that we receive changes for that would cause annotations (e.g. I don't have an editor open for a file another session participant is currently editing). (This also covers the case that the editor for such a resource is closed and re-opened during a session.)

Such annotations should still be tracked by Saros. To do so, they are added to the annotation store without an editor. As a result, these annotations are only stored as offsets and not shown anywhere (at this point). As soon as the user open an editor for the resource, the editor will be added to all stored annotations and they will be added to that editor (which is called "adding a local representation" in the annotation manager logic). This happens in [AnnotationUpdater](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/eventhandler/editor/editorstate/AnnotationUpdater.java#L21).